blog.md

```diff
@@ -0,0 +1,258 @@
+Buckets may store key/value pairs in bytes and sub-Buckets. 
+Sub-buckets may be stored inline inside a parent bucket or on its own page.  
+
+The data types used to represent the BBolt database are relatively straightforward. 
+
+```rust
+/// The Page ID. Page address = `PgId` * meta.page_size
+#[repr(C)]
+pub struct PgId(u64);
+
+/// The Transaction ID. Monotonic and incremented every commit
+#[repr(C)]
+pub struct TxId(u64);
+
+/// `PageHeader` represents the on-file layout of a page header. 
+/// 
+/// `page` in Go BBolt
+#[repr(C)]
+pub struct PageHeader {
+  /// This Page's ID
+  id: PgId,
+  /// Page's type. Branch(0x01), Leaf(0x02), Meta(0x04), or FreeList(0x10)
+  flags: u16,
+  /// Defines the number of items in the Branch, Leaf, and Freelist pages
+  count: u16,
+  /// How many additional meta.page_size pages are included in this page
+  overflow: u32,
+}
+
+/// `BucketHeader` represents the on-file layout of a bucket header.
+/// This is stored as the "value" of a bucket key. If the bucket is small enough,
+/// then its root page can be stored inline in the "value", after the bucket
+/// header. In the case of inline buckets, the "root" will be 0.
+/// 
+/// `bucket` in Go BBolt
+#[repr(C)]
+pub struct BucketHeader {
+  /// Page ID of the bucket's root-level page
+  root: PgId,
+  /// monotonically incrementing, used by NextSequence()
+  sequence: u64,
+}
+
+/// `Meta` represents the on-file layout of a database's metadata
+/// 
+/// `meta` in Go BBolt
+#[repr(C)]
+pub struct Meta {
+  /// Uniquely ID for BBolt databases 
+  magic: u32,
+  /// Database version number
+  version: u32,
+  /// Database page size where page address = [PgId] * meta.page_size
+  page_size: u32,
+  flags: u32,
+  /// Root bucket header
+  root: BucketHeader,
+  /// FreeList page location
+  free_list: PgId,
+  /// The end of the database where EOF = meta.pgid * meta.page_size
+  pgid: PgId,
+  /// Current transaction ID
+  txid: TxId,
+  /// Checksum of the previous Meta fields using the 64-bit version of the Fowler-Noll-Vo hash function
+  checksum: u64,
+}
+
+/// `LeafPageElement` represents the on-file layout of a leaf page's element
+/// 
+/// `leafPageElement` in Go BBolt 
+#[repr(C)]
+pub struct LeafPageElement {
+  /// Additional flag for each element. If leaf is a Bucket then 0x01 set
+  flags: u32,
+  /// The distance from this element's pointer to its key/value location
+  pos: u32,
+  /// Key length
+  key_size: u32,
+  /// Value length
+  value_size: u32,
+}
+
+///`BranchPageElement` represents the on-file layout of a branch page's element
+/// 
+/// `branchPageElement` in Go BBolt
+#[repr(C)]
+pub struct BranchPageElement {
+  /// The distance from this element's pointer to its key location
+  pos: u32,
+  /// Key length
+  key_size: u32,
+  /// Page ID of this branch's
+  pgid: PgId,
+}
+
+```
+
+The 4 database page types are also straightforward. 
+
+Page Layouts
+```
+      Meta Page
```

Comment:
Similar to the codeblock before, break each of these (really nice) diagrams into separate codeblocks and provide a blurb for each, with transitions/captions.

The goal is to be able to skim the article and get an idea of what's contained therein. Naturally, if I see largely two big codeblocks with very little English, I'm inclined to just assume it's a dump of some code and lose interest.